File: storage/state.py
from dataclasses import dataclass, field
from typing import Dict, Optional

from merkle import MerkleTree
from utils import h_join

@dataclass
class TimeStateTree:
    """表示单个文件状态随时间变化的默克尔树。"""
    leaves: Dict[int, str] = field(default_factory=dict)
    merkle: Optional[MerkleTree] = None

    def build(self):
        max_index = max(self.leaves.keys()) if self.leaves else -1
        seq = [self.leaves.get(i, h_join("missing", str(i))) for i in range(max_index + 1)]
        self.merkle = MerkleTree(seq)

    def root(self) -> str:
        return self.merkle.root() if self.merkle else h_join("empty")

@dataclass
class StorageStateTree:
    """聚合单个节点所有TimeStateTree根的默克尔树。"""
    file_roots: Dict[str, str] = field(default_factory=dict)
    merkle: Optional[MerkleTree] = None

    def build(self):
        leaves = [self.file_roots[fid] for fid in sorted(self.file_roots.keys())]
        self.merkle = MerkleTree(leaves)

    def root(self) -> str:
        return self.merkle.root() if self.merkle else h_join("empty")

@dataclass
class ServerStorage:
    """服务器节点整个存储状态的容器。"""
    time_trees: Dict[str, TimeStateTree] = field(default_factory=dict)
    storage_tree: Optional[StorageStateTree] = None

    def add_chunk_commitment(self, file_id: str, index: int, commitment: str):
        self.time_trees.setdefault(file_id, TimeStateTree()).leaves[index] = commitment

    def build_state(self):
        for tst in self.time_trees.values():
            tst.build()
        self.storage_tree = StorageStateTree({fid: tst.root() for fid, tst in self.time_trees.items()})
        self.storage_tree.build()

    def storage_root(self) -> str:
        return self.storage_tree.root() if self.storage_tree else h_join("empty")

    def num_files(self) -> int:
        return len(self.time_trees)

File: src/storage/state.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use crate::merkle::MerkleTree;
use crate::utils::h_join;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TimeStateTree {
    pub leaves: HashMap<usize, String>,
    #[serde(skip)]
    pub merkle: Option<MerkleTree>,
}

impl Default for TimeStateTree {
    fn default() -> Self {
        Self {
            leaves: HashMap::new(),
            merkle: None,
        }
    }
}

impl TimeStateTree {
    pub fn build(&mut self) {
        if self.leaves.is_empty() {
            self.merkle = Some(MerkleTree::new(&[]));
            return;
        }
        let max_index = self.leaves.keys().copied().max().unwrap();
        let seq: Vec<String> = (0..=max_index)
            .map(|i| {
                self.leaves
                    .get(&i)
                    .cloned()
                    .unwrap_or_else(|| h_join(["missing", &i.to_string()]))
            })
            .collect();
        self.merkle = Some(MerkleTree::new(&seq));
    }

    pub fn root(&self) -> String {
        self.merkle
            .as_ref()
            .map(|m| m.root())
            .unwrap_or_else(|| h_join(["empty"]))
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StorageStateTree {
    pub file_roots: HashMap<String, String>,
    #[serde(skip)]
    pub merkle: Option<MerkleTree>,
}

impl Default for StorageStateTree {
    fn default() -> Self {
        Self {
            file_roots: HashMap::new(),
            merkle: None,
        }
    }
}

impl StorageStateTree {
    pub fn build(&mut self) {
        let mut entries: Vec<(String, String)> = self
            .file_roots
            .iter()
            .map(|(fid, root)| (fid.clone(), root.clone()))
            .collect();
        entries.sort_by(|a, b| a.0.cmp(&b.0));
        let seq: Vec<String> = entries.into_iter().map(|(_, root)| root).collect();
        self.merkle = Some(MerkleTree::new(&seq));
    }

    pub fn root(&self) -> String {
        self.merkle
            .as_ref()
            .map(|m| m.root())
            .unwrap_or_else(|| h_join(["empty"]))
    }
}

#[derive(Debug, Default)]
pub struct ServerStorage {
    pub time_trees: HashMap<String, TimeStateTree>,
    pub storage_tree: Option<StorageStateTree>,
}

impl ServerStorage {
    pub fn add_chunk_commitment(&mut self, file_id: &str, index: usize, commitment: String) {
        self.time_trees
            .entry(file_id.to_string())
            .or_default()
            .leaves
            .insert(index, commitment);
    }

    pub fn build_state(&mut self) {
        for tst in self.time_trees.values_mut() {
            tst.build();
        }
        let mut file_roots = HashMap::new();
        for (fid, tst) in &self.time_trees {
            file_roots.insert(fid.clone(), tst.root());
        }
        let mut storage_tree = StorageStateTree {
            file_roots,
            merkle: None,
        };
        storage_tree.build();
        self.storage_tree = Some(storage_tree);
    }

    pub fn storage_root(&self) -> String {
        self.storage_tree
            .as_ref()
            .map(|st| st.root())
            .unwrap_or_else(|| h_join(["empty"]))
    }

    pub fn num_files(&self) -> usize {
        self.time_trees.len()
    }
}
